### PR TITLE
Implement basic graceful drain/exit of proxy

### PR DESCRIPTION
**Pull request by howardjohn**
Currently the proxy just runs forever on SIGTERM. This makes it exit, and puts some *very basic* graceful draining in place.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -46,6 +46,7 @@ hyper-boring = { version= "2.1.2", features = ['fips'] }
 anyhow = "1.0.65"
 serde_yaml = "0.9.13"
 async-trait = "0.1.58"
+drain = "0.1.1"
```
Comment (stevenctl) [inline Cargo.toml:49]: I see this is from linkerd and is probably solid. 

As we add more reviewers (not sure how many will be project-root approvers), we gotta have a well-defined dependency evaluation process.

File: src/proxy/mod.rs [hunk 2]
```diff
@@ -27,11 +28,12 @@ impl Proxy {
         cfg: config::Config,
         workloads: WorkloadInformation,
         secret_manager: identity::SecretManager,
+        drain: Watch,
     ) -> Result<Proxy, Error> {
         // We setup all the listeners first so we can capture any errors that should block startup
         let inbound_passthrough = InboundPassthrough::new(cfg.clone());
         let inbound = Inbound::new(cfg.clone(), workloads.clone(), secret_manager.clone()).await?;
-        let outbound = Outbound::new(cfg.clone(), secret_manager, workloads).await?;
+        let outbound = Outbound::new(cfg.clone(), secret_manager, workloads, drain).await?;
```
Comment (stevenctl) [inline src/proxy/mod.rs:36]: at some point as we keep passing the `drain` down, does it make sense to clone it (even though there aren't other uses)

makes it clear that each consumer should take a clone. i think a codebase for a language we're not that familiar with it's nice when you're surrounded by examples of how you can get the same functionality. 

random thought, not sure if it actually makes much sense. the clone should be nearly free.

**Review by stevenctl — COMMENTED** [review]
LGTM. Can we add any tests here. We should probably stop merging untested PRs soon.

File: src/proxy/mod.rs (hunk 2 above)
Comment (howardjohn) [reply src/proxy/mod.rs:36]: Yeah, probably so

**Comment by howardjohn** [comment]
> LGTM. Can we add any tests here. We should probably stop merging untested PRs soon.

I agree. This one will be a good test of our test setup, since it involves the whole main() structure, so will be a ~full e2e test.

**Review by stevenctl — APPROVED** [review]

**Pull request merged**